### ISSUE TITLE
Maintained specification for posts/votes/etc

### ISSUE THREAD

**Issue by christianbundy** [fraction/oasis#278]
**What's the problem you want solved?** It's hard to understand what specification we're following when we read and write posts because it isn't written down anywhere in this repo. Right now we're just going off of a subjective feeling about what's expected in different clients, but that feels bad.

**Is there a solution you'd like to recommend?** Either we iteratively write the spec ourselves, covering the surface area that we're supporting, or we should link to another URL that's actually maintained by a person with a name.

**Comment by mixmix** [comment]
this issue is on the oasis repo.
Let's start `ssbc/thread-spec` and move this issue there

**Comment by christianbundy** [comment]
> this issue is on the oasis repo.

Yep, that's intentional. I think it would be helpful for beginners to have a direct link like "the spec for posts is available at $X". I think this issue could be resolved by either making an ssbc/thread-spec or ssbc/ssb-spec-drafts or wherever and fleshing it out and then linking there *or* just writing it as a doc in this repo and forking whenever it becomes a pain point.

Either way works fine for me, it's just important to me that the "where is the spec" problem is resolved in this repo.

**Comment by mixmix** [comment]
Sounds reasonable

The thread spec will take a few weeks to do.
I might be able to do it async.... but we need to have clear buy in from some people about pushing till it's done (/ done enough).
Could the first step be to get in a call / shared document / thread and list all the things which a thread should / shouldn't do as a starting point?

**Comment by mixmix** [comment]
mmm, yeah I guess I'm asking **when** do we do this, and **who's in**?

My picks would be : @christianbundy @mixmix @soapdog (and @arj / @arj03 if they're interested )

**Comment by soapdog** [comment]
I'm interested but I'm biased. I like SSB msgs that don't require much querying to reduce them to a useful state as I'm pumping all my stuff through `ssb-ws` and am not in control of the backend. So if we could simply agree on how `branch` and `fork` should work for posts, I'd be happy.

**Comment by christianbundy** [comment]
Quick brain dump of what we have in Oasis IIRC.

Mandatory:

```js
{
  type: 'post',
  text: 'ssb-flavored markdown goes here'
}
```

Optional channel, like a hashtag embedded in JSON: 

```js
{
  channel: '#30charsmaxIthink'
}
```

Optional reply to thread:

```js
{
  root: '%thread root', 
  branch: ['%unreferenced.with.same.root']}`
}
```

Optional reply to message

```js
{
  root: '%message you're replying to',
  fork: '%previous thread root'
}
```

Optional mentions: 

```js
{
  mentions: ['@abc']
}
```

or

```js
{
  mentions: [
    {
      name: 'aardvark',
      link: '@abc'
    }
  ]
}
```

**Comment by soapdog** [comment]
@christianbundy AFAIK the `channel` property does not contain the hash, so if the channel is #patchfox, then `channel = "patchfox"`.

**Comment by cinnamon-bun** [comment]
Yes please!!  Decentralization doesn't work very well when it comes to documentation ;)

I know of these links, but some are old.
https://github.com/ssbc/ssb-spec-drafts
https://github.com/ssbc/is-my-ssb-valid
https://github.com/ssbc/ssb-typescript
https://github.com/ssb-junkyard/ssb-msg-schemas

**Comment by mixmix** [comment]
also : https://github.com/ssbc/ssb-schema-definitions


Agree with all of that except that "reply to message" @christianbundy :+1: 

```
{
  root: '%message you're replying to',
  fork: '%previous thread root'
}
```

Things I want to pin down:
- whether this message is considered part of the thread (nested reply), or a fork?
    - i.e. which tangle(s) is this message a part of?
- do we want to support multiple forks/ nests of the same message, if so how?
- what does a reply to one of these fork/ nest messages look like, what fields does it have?

**Comment by christianbundy** [comment]
> whether this message is considered part of the thread (nested reply), or a fork? 

I think of the thread being the set of messages that share the same `root`, plus the root message that they refer to. Patchwork called them forks but also displayed links to them in the thread, so I think the answer to your question is "yes". I don't think that the post specification can disambiguate between "small aside" and "big aside", although I agree that it's a presentation concern.

> i.e. which tangle(s) is this message a part of?

I think it would be consistent for the `branch` property to be set based on the `root` property. **Really**, I think we should be tangling *all* messages instead of just messages in a thread, but that's a whole other can of worms. I think 

> do we want to support multiple forks/ nests of the same message, if so how?

Yes, so if creating a fork/nest is done with:

```js
{
  root: '%message you're replying to',
  fork: '%previous thread root'
}
```

> what does a reply to one of these fork/ nest messages look like, what fields does it have?

In Oasis we just do:

```js
{
  root: '%fork/nest message'
}
```

Here's a diagram explaining it in detail:

![graph](https://user-images.githubusercontent.com/537700/75721828-e330ed00-5c8d-11ea-841b-14fe9171cbc7.png)

Replies to fork/nest messages *could* have a `fork` that point two roots up (to `C`) for ultimate consistency, but I'm not sure what everyone's preferences are.

**Comment by cinnamon-bun** [comment]
Also dropping this link here from 10 months ago, a big thread about Compatibility Testing between clients.  Mostly focused on low level protocol things though.

%F0ePtbVPwWY5PaT77V6mKwvkQqgXh5EnHqHtIxeBbWo=.sha256

**Comment by mixmix** [comment]
@christianbundy this confused me a lot until I realised that you'd drawn time from bottom to top!

Things to note:
1. :+1: this makes a new type of message to test for "a thread start" which can be either:
    - `root: null` (no fork) :  this is the ultimate root/ OG root?
    - `root: %some-msg` + `fork:' %OG-root'` : this is a fork root
2. because of (1) we can have multiple unique, distinguishable forks of same point
3. you've identified this unique type in (1) as the "ID" for the thread. (e.g. "thead m")
    - I recently realised the `root` tangle attribute is actually the "tangle-id"
4. we are not clear on whether forks / nests are the same thing
    - to me what you've drawn in a "fork", because with a single backlinks lookup on msg `A`, you'll get all of "thread A", and learn about the existence of fork-roots `m` + `x` (but NOT the content of those fork threads)
    - I want to build forks and leave space for "nests" which I see as _in thread asides_. I'm not sure why this is important to me. It may not be important once we have actual forks.
5. we need to draw a fork off a fork (or a nest of a nest)

**Comment by mixmix** [comment]
Things this diagram doesn't yet cover: 
6. what are the `branch` fields for each msg? (which I've been calling `previous` in new models)
7. what do we want to support i.t.o. concurrent posting.
    - a) do we want to force linearity (as we currently do) and ignore the information that people aren't speaking into the same space always?
    - b) do we want to enable clients which reveal hydras and merges in the tangle?
